numba/np/arraymath.py

```diff
@@ -665,6 +666,136 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@register_jitable
+def _impl_iscomplex_isreal_arr(fn, x):
+    ax = np.asarray(x) # NumPy uses asanyarray here!
+    return fn(ax.imag, 0)
+
+
+@register_jitable
+def np_iscomplex_isreal(x, operator_fn, opt_ret):
+
+    if type_can_asarray(x):
+        return lambda x: _impl_iscomplex_isreal_arr(operator_fn, x)
+    elif isinstance(x, types.Optional):
+        def impl(x):
+            if x is None:
+                return opt_ret
+            return _impl_iscomplex_isreal_arr(operator_fn, x)
+    else:
+        def impl(x):
+            return False
```

Comment:
Should this path just be `return None`, so that it's a compilation error?

Reply:
To be a compilation error, I guess one must explicitly raise an exception. For this particular case, I think we should put the exception inside the `if` condition for `types.Optional`. There are cases where the input is "strange" and NumPy gives an output. For instance:

```python
In [51]: np.iscomplex(None)
Out[51]: False

In [52]: np.iscomplex("hello")
Out[52]: False

In [53]: np.isreal(None)
Out[53]: True

In [54]: np.isreal('Hello')
Out[54]: False
```

Reply:
I thought it was sufficient to just not return a `impl` from the overload?

Reply:
I'm not sure about this either. If returning `None` is enough.

Reply:
Returning `None` from `iscomplexobj` returns 

```python
from numba import njit
import numpy as np

@njit
def foo(x):
    optional = np.array(12j)
    if x > 0:
        optional = None

    return np.iscomplexobj(optional)

print(foo(1))
print(foo(-1))
```

```bash
$ python3 file.py
Traceback (most recent call last):
  File "a.py", line 12, in <module>
    print(foo(1))
TypeError: expected array(complex128, 0d, C), got None
(numba)
```

I would change this to manually raise a `TypingError` with the following message:
`TypingError: Expected array(complex128, 0d, C) but got Optional[array(complex128, 0d, C), None]`

```
$ python3 file.py
Traceback (most recent call last):
  File "a.py", line 12, in <module>
    print(foo(1))
  File "/Users/guilhermeleobas/Documents/GitHub/numba/numba/core/dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/guilhermeleobas/Documents/GitHub/numba/numba/core/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "/Users/guilhermeleobas/Documents/GitHub/numba/numba/core/utils.py", line 77, in reraise
    raise value.with_traceback(tb)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function iscomplexobj at 0x10ea425f0>) with argument(s) of type(s): (OptionalType(array(complex128, 0d, C)) i.e. the type 'array(complex128, 0d, C) or None')
 * parameterized
In definition 0:
    TypingError: Expected array(complex128, 0d, C) but got Optional[array(complex128, 0d, C), None]
    raised from /Users/guilhermeleobas/Documents/GitHub/numba/numba/np/arraymath.py:709
In definition 1:
    TypingError: Expected array(complex128, 0d, C) but got Optional[array(complex128, 0d, C), None]
    raised from /Users/guilhermeleobas/Documents/GitHub/numba/numba/np/arraymath.py:709
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function iscomplexobj at 0x10ea425f0>)
[2] During: typing of call at a.py (10)


File "a.py", line 10:
def foo(x):
    <source elided>

    return np.iscomplexobj(optional)
    ^

(numba)
```